Changing webhook in remote to use the right image

MbolotSuse opened the pull request:
## Issue:  
#40685

## Problem
The rancher webhook is now deployed in downstream clusters. This caused an issue where, in airgap scenarios, the webhook and the pods that were deploying the webhook would not use the specified airgap registry. This change allows the agent to pass down an override image, when applicable.
 
## Solution
The systemcharts install logic now accepts an override image, which is passed down by the agent. 

In the case where an override image was specified for the webhook, the systemDefaulRegistry value for the webhook's chart will not be set. In addition, the override image will be specified using the `values.image.repository` field. 

In addition, the chart.Manager (used by system charts for installing/upgrading charts) now accepts a parameter for an override image. All other consumers of this logic (fleet, csp adapter) pass down the empty string in all cases.
 
## Testing
Repro steps are valid.

## Engineering Testing
### Manual Testing

- I simulated a systemDefaultRegistry change by changing this value to `docker.io` and validating that the full path of the image (e.x. `docker.io/rancher/webhook`) was used. 
- I setup a downstream cluster proxy (used docker registry image and nginx as a proxy auth) and validated that the webhook/shell image used the specified cluster registry and were able to pull images that required authentication.

## QA Testing Considerations
- Webhook install in the local cluster 
- Webhook install in local cluster with system default registry set
- Webhook install in the remote cluster
- Webhook install in the remote cluster with system default registry set
- Webhook install in the remote cluster with cluster specific registry set (un-authenticated)
- Webhook install in the remote cluster with cluster specific registry set (authenticated)
- Webhook install in the remote cluster with cluster specific registry set (un-authenticated) and system default registry set
- Webhook install in the remote cluster with cluster specific registry set (authenticated) and system default registry set
 
 In addition, this should be tested with various cluster types (specifically for the authenticated registry case), including:
 - RKE
 - RKE2
 - K3s
 - Hosted providers (EKS, AKS, GKE)
 
 This last set of tests (for different types of clusters) is due to how the authenticated registries work. On RKE/RKE2/K3s, the secrets to pull images appear to be set at a cluster/container runtime level, meaning that the webhook/shell don't need to specify imagePullSecrets manually (thanks to @jakefhyde for looking into this from team 2's side). However, it would be a good idea to test if this is the case for all types of clusters (note: if this is broke, it's likely that chart functionality also has issues on that setup). 
 
### Regressions Considerations
- System charts (fleet, csp-adapter, webhook) installation/ugprade

MbolotSuse commented:
I still need to test this more with the local cluster and authenticated downstream cluster repos, so marking this as a WIP for now until I get those setups going.

KevinJoiner commented on pkg/controllers/dashboard/systemcharts/controller.go:
```diff
@@ -20,13 +20,19 @@ import (
 )
 
-func Register(ctx context.Context, wContext *wrangler.Context) error {
+type ImageOverride struct {
+	Registry string
+}
```
Do we need a struct for this?

MbolotSuse replied:
The idea was to future proof this for the addition of ImagePullSecrets (for authenticated registries). I'm testing now to see if that's necessary. If it's not, I'll remove this and set it as a string instead.

MbolotSuse replied:
I've removed the struct and instead moved to a string - Authenticated registries appear to be set at a container-runtime/cluster level, so we don't seem to need specific image pull secrets here.

rmweir reviewed (approved):
LGTM

MbolotSuse commented:
I've pushed some changes, namely:
- Since we don't need to pass down image-pull secrets (see https://github.com/rancher/rancher/pull/40726#discussion_r1122154351), I've removed the struct in favor of a basic string
- The systemDefaultRegistry map can't be totally unset. There's parts in the chart (namely this part: https://github.com/rancher/webhook/blob/b99015e507e2d9d63c93feabf6fd9f11fac127a6/charts/rancher-webhook/templates/deployment.yaml#L20) that require the `global : {systemDefaultRegistry: ""}` to still be set, even if it's nothing. Because of this, I've adjusted this to just set the default registry to "" in case an override image is passed in
- I added a basic test case for the image override case inside of the webhook's deploy logic.
